servant/src/Servant/API/Cookie.hs

```diff
@@ -0,0 +1,23 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveFunctor      #-}
+{-# LANGUAGE PolyKinds          #-}
+{-# OPTIONS_HADDOCK not-home    #-}
+module Servant.API.Cookie where
+
+import Data.ByteString    (ByteString)
+import Data.Typeable      (Typeable)
+import GHC.TypeLits       (Symbol)
+
+-- | Extract the given cookie's value as a value of type @a@.
+--
+-- Example:
+--
+-- >>>            -- GET /test
+-- >>> type MyApi = "test" :> Cookie "token" Text :> Get '[JSON] Text
+
+data Cookie (sym :: Symbol) a = Cookie a
+                              | MissingCookie
```

Comment:
Yeah, in `Header` it's used for setting headers more easily, but since I don't think we should worry about setting cookies as part of that PR, I'd leave it out for now.